File: leetcode/00200_number_of_islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        rows = len(grid)
        cols = len(grid[0]) if grid else 0    
        visited = [[False] * cols for _ in range(rows)]
        di_dj = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        def visit(i: int, j: int) -> None:
            visited[i][j] = True
            for di, dj in di_dj:
                ni, nj = i + di, j + dj
                if ni < 0 or ni >= rows:
                    continue
                if nj < 0 or nj >= cols:
                    continue
                if visited[ni][nj] or grid[ni][nj] == '0':    
                    continue
                visit(ni, nj)
        answer = 0
        for i in range(rows):
            for j in range(cols):
                if grid[i][j] == '1' and not visited[i][j]:
                    visit(i, j)
                    answer += 1
        return answer

File: leetcode/00200_number_of_islands.rs
use std::collections::VecDeque;

impl Solution {
    pub fn num_islands(mut grid: Vec<Vec<char>>) -> i32 {
        let (rows, cols) = (grid.len(), grid[0].len());
        let mut ans = 0;
        let mut q = VecDeque::<(usize, usize)>::new();

        for i in 0..rows {
            for j in 0..cols {
                if grid[i][j] != '1' { continue; }
                ans += 1;
                q.push_back((i, j));
                while !q.is_empty() {
                    let (r, c) = q.pop_front().unwrap();
                    if grid[r][c] == '*' { continue; }
                    grid[r][c] = '*';
                    if r > 0 && grid[r - 1][c] == '1' {
                        q.push_back((r - 1, c));
                    }
                    if r < rows - 1 && grid[r + 1][c] == '1' {
                        q.push_back((r + 1, c));
                    }
                    if c > 0 && grid[r][c - 1] == '1' {
                        q.push_back((r, c - 1));
                    }
                    if c < cols - 1 && grid[r][c + 1] == '1' {
                        q.push_back((r, c + 1));
                    }
                }
            }
        }
        ans
    }
}

